src/runtime/particle.ts

```diff
@@ -92,8 +106,22 @@ export class Particle {
   protected async setHandles(handles: ReadonlyMap<string, Handle>): Promise<void> {
   }
 
+  private _countInputHandles(handles: ReadonlyMap<string, Handle>): number {
+    let count = 0;
+    for (const [name, handle] of handles) {
+      if (handle.canRead) {
+        count++;
+      }
+    }
+    return count;
+  }
+
   async callOnHandleSync(handle: Handle, model, onException: Consumer<Error>) {
     await this.invokeSafely(async p => p.onHandleSync(handle, model), onException);
+    // once we've synced each readable handle, we are ready to start
+    if (--this._handlesToSync === 0) {
```

Comment:
It would be safer to use a set of the handles being synced rather than just counting. While it's extremely unlikely, it's possible for one handle to sync twice before all other handles are synced.